### PR TITLE
add https support for wrapper integration

### PR DESCRIPTION
**Pull request by alberts-tid**
add Https suppport for wrapper integration

@flopezag @jframos  @jesuspg

### CODE REVIEW

**Comment by flopezag** [comment]
Coverage decreased (-1.48%)...

File: core/src/main/java/com/telefonica/euro_iaas/sdc/installator/impl/InstallatorPuppetImpl.java [hunk 1]
```diff
@@ -80,7 +84,7 @@ public void callService(VM vm, String vdc, ProductRelease product, String action
         } catch (InstallatorException e) {
             log.warn("It is not possible to generate the manifests in the puppet master " + e.getMessage());
             throw new InstallatorException(e.getMessage());
-        }
+        }
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/installator/impl/InstallatorPuppetImpl.java]: ?

File: core/src/main/java/com/telefonica/euro_iaas/sdc/installator/impl/InstallatorPuppetImpl.java [hunk 2]
```diff
@@ -110,7 +118,11 @@ public void generateFilesinPuppetMaster(VM vm, String vdc, ProductRelease produc
     public void callService(ProductInstance productInstance, VM vm, List<Attribute> attributes, String action,
             String token) throws InstallatorException, NodeExecutionException {
 
-        callPuppetMaster(vm, productInstance.getVdc(), productInstance.getProductRelease(), action, token, attributes);
+        try {
+            callPuppetMaster(vm, productInstance.getVdc(), productInstance.getProductRelease(), action, token, attributes);
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/installator/impl/InstallatorPuppetImpl.java]: gt 120

File: core/src/main/java/com/telefonica/euro_iaas/sdc/installator/impl/InstallatorPuppetImpl.java [hunk 3]
```diff
@@ -187,10 +203,7 @@ private void callPuppetMaster(VM vm, String vdc, ProductRelease product, String
         log.info("connecting to puppetURL: " + "puppetURL: " + puppetUrl + "v2/node/" + vm.getHostname() + "/" + action);
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/installator/impl/InstallatorPuppetImpl.java]: gt 120

File: core/src/main/java/com/telefonica/euro_iaas/sdc/installator/impl/InstallatorPuppetImpl.java [hunk 4]
```diff
@@ -187,10 +203,7 @@ private void callPuppetMaster(VM vm, String vdc, ProductRelease product, String
...
         log.info("payload: " + payload);
         try {
-            response = client.execute(postInstall);
-            int statusCode = response.getStatusLine().getStatusCode();
-            HttpEntity entity = response.getEntity();
-            EntityUtils.consume(entity);
+            int statusCode = httpsClient.doHttpsPost(puppetUrl + "v2/node/" + vm.getHostname() + "/" + action, payload,headers);
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/installator/impl/InstallatorPuppetImpl.java]: gt 120

File: core/src/main/java/com/telefonica/euro_iaas/sdc/installator/impl/InstallatorPuppetImpl.java [hunk 5]
```diff
@@ -226,7 +230,7 @@ private void callPuppetMaster(VM vm, String vdc, ProductRelease product, String
         } catch (IllegalStateException e1) {
             log.error(e1.getMessage());
             throw new InstallatorException(e1);
-        }
+        }
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/installator/impl/InstallatorPuppetImpl.java]: While space at the end?

File: core/src/main/java/com/telefonica/euro_iaas/sdc/installator/impl/InstallatorPuppetImpl.java [hunk 6]
```diff
@@ -226,7 +230,7 @@ private void callPuppetMaster(VM vm, String vdc, ProductRelease product, String
         } catch (IllegalStateException e1) {
             log.error(e1.getMessage());
             throw new InstallatorException(e1);
-        }
+        } 
     }
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/installator/impl/InstallatorPuppetImpl.java]: I think that you should explain here the functionality of this method, mainly to know why do you need to substract 2 to the length of the newValue substring.

File: core/src/main/java/com/telefonica/euro_iaas/sdc/manager/impl/NodeManagerImpl.java [hunk 7]
```diff
@@ -111,29 +117,26 @@ private void puppetDelete(String vdc, String nodeName, String token) throws Node
 
         puppetLog.info("deleting node " + nodeName + " from puppet master");
 
-        HttpDelete delete = null;
+        String deleteUrl = null;
         try {
-            delete = new HttpDelete(openStackRegion.getPuppetWrapperEndPoint(token) + "v2/node/" + nodeName);
+            deleteUrl = openStackRegion.getPuppetWrapperEndPoint(token) + "v2/node/" + nodeName;
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/manager/impl/NodeManagerImpl.java]: It is mean that you have one puppet instance per each region. Is it correct?

is it the same for Chef?

File: core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+import javax.net.ssl.X509TrustManager;
+
+import org.hibernate.context.TenantIdentifierMismatchException;
+
+public class HttpsClient {
+
+    public static final String HEADER_AUTH = "X-Auth-Token";
+    public static final String HEADER_TENNANT = "Tenant-Id";
+    
+    private HttpsURLConnection connectionSetup(URL url) throws NoSuchAlgorithmException, KeyManagementException, IOException{
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java]: gt 120

File: core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+            BufferedReader br = new BufferedReader(new InputStreamReader(con.getInputStream()));
+            String line;
+
+            while ((line = br.readLine()) != null) {
+                System.out.println(line);
+                sb.append(line);
+            }
+            br.close();
+        }
+        int responseStatus = con.getResponseCode();
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java]: why do you not move this line before the condition, currently you are taking twice the value from con.getResponseCode()

File: core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        con.disconnect();
+        return responseStatus;
+    }
+
+    
+    public int doHttpsGet(String _url, String payload, Map<String, String> headers) throws KeyManagementException, NoSuchAlgorithmException, IOException {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java]: gt 120

File: core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+            BufferedReader br = new BufferedReader(new InputStreamReader(con.getInputStream()));
+            String line;
+
+            while ((line = br.readLine()) != null) {
+                System.out.println(line);
+                sb.append(line);
+            }
+            br.close();
+        }
+        int responseStatus = con.getResponseCode();
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java]: Idem line 106

File: core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+        con.disconnect();
+        return responseStatus;
+    }
+    
+    public int doHttpsDelete(String _url, String payload, Map<String, String> headers) throws KeyManagementException, NoSuchAlgorithmException, IOException {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java]: gt 120

File: core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+        URL url = new URL(_url);
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java]: Well,

You have done lots of copy/paste in this code, you should refactoring the code of all those methos due to the content is the same, Except the param DELETE, GET, POST.

Why do you not create a method that should be invoked with the appropriate param...

File: core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java [hunk 14]
```diff
@@ -64,11 +66,13 @@
 import com.telefonica.euro_iaas.sdc.model.dto.AttributeDto;
 import com.telefonica.euro_iaas.sdc.model.dto.PuppetNode;
 import com.telefonica.euro_iaas.sdc.model.dto.VM;
+import com.telefonica.euro_iaas.sdc.util.HttpsClient;
 import com.telefonica.euro_iaas.sdc.util.SystemPropertiesProvider;
```
Comment (flopezag) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java]: Please provide a description of the class and of all the methods (at least public ones).

File: core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java [hunk 15]
```diff
@@ -137,35 +141,39 @@ public void setup() throws ClientProtocolException, IOException, OpenStackExcept
...
         puppetInstallator.setClient(client);
         puppetInstallator.setOpenStackRegion(openStackRegion);
+        puppetInstallator.setHttpsClient(httpsClient);
         when(openStackRegion.getPuppetWrapperEndPoint("token")).thenReturn("http://");
 
     }
 
     @Test
-    public void testGenerateFilesinPuppetMaster() throws InstallatorException, NodeExecutionException {
+    public void testGenerateFilesinPuppetMaster() throws InstallatorException, NodeExecutionException, KeyManagementException, NoSuchAlgorithmException, IOException {
```
Comment (flopezag) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java]: gt 120

File: core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java [hunk 16]
```diff
@@ -137,35 +141,39 @@ public void setup() throws ClientProtocolException, IOException, OpenStackExcept
...
 
         when(statusLine.getStatusCode()).thenReturn(200);
+        when(httpsClient.doHttpsPost(Mockito.anyString(), Mockito.anyString(),(Map<String,String>)Mockito.anyObject())).thenReturn(200);
```
Comment (flopezag) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java]: gt 120

File: core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java [hunk 17]
```diff
@@ -137,35 +141,39 @@ public void setup() throws ClientProtocolException, IOException, OpenStackExcept
...
         when(statusLine.getStatusCode()).thenReturn(200);
         puppetInstallator.callService(host, "test", productRelease, "install", "token");
     }
 
     @Test(expected = InstallatorException.class)
-    public void testCallService_FAIL() throws InstallatorException, NodeExecutionException, OpenStackException {
+    public void testCallService_FAIL() throws Exception {
 
-        when(statusLine.getStatusCode()).thenReturn(500);
+        when(httpsClient.doHttpsPost(Mockito.anyString(), Mockito.anyString(),(Map<String,String>)Mockito.anyObject())).thenReturn(500);
```
Comment (flopezag) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java]: gt 120

File: core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java [hunk 18]
```diff
@@ -137,35 +141,39 @@ public void setup() throws ClientProtocolException, IOException, OpenStackExcept
...
         puppetInstallator.generateFilesinPuppetMaster(host, "test", productRelease, "install", "token");
         puppetInstallator.callService(host, "test", productRelease, "install", "token");
     }
 
     @Test(expected = InstallatorException.class)
-    public void testCallService_1_OK_1_FAIL() throws InstallatorException, NodeExecutionException {
+    public void testCallService_1_OK_1_FAIL() throws InstallatorException, NodeExecutionException, KeyManagementException, NoSuchAlgorithmException, IOException {
```
Comment (flopezag) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java]: gt 120

File: core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java [hunk 19]
```diff
@@ -241,9 +249,11 @@ public void testLoadNodeNoExists() throws OpenStackException, CanNotCallPuppetEx
     }
 
     @Test
-    public void testCallService_attributes_all_OK() throws InstallatorException, NodeExecutionException, IOException {
+    public void testCallService_attributes_all_OK() throws InstallatorException, NodeExecutionException, IOException, KeyManagementException, NoSuchAlgorithmException {
```
Comment (flopezag) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java]: gt 120

File: core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java [hunk 20]
```diff
@@ -241,9 +249,11 @@ public void testLoadNodeNoExists() throws OpenStackException, CanNotCallPuppetEx
...
 
-        when(statusLine.getStatusCode()).thenReturn(200);
+//        when(statusLine.getStatusCode()).thenReturn(200);
+        when(httpsClient.doHttpsPost(Mockito.anyString(), Mockito.anyString(),(Map<String,String>)Mockito.anyObject())).thenReturn(200);
```
Comment (flopezag) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java]: gt 120

File: core/src/test/java/com/telefonica/euro_iaas/sdc/manager/NodeManagerImplTest.java [hunk 21]
```diff
@@ -89,7 +97,7 @@ public void setup() throws CanNotCallChefException, OpenStackException {
     }
 
     @Test
-    public void deleteNodeTestOK() throws NodeExecutionException, CanNotCallChefException, EntityNotFoundException, ClientProtocolException, IOException {
+    public void deleteNodeTestOK() throws NodeExecutionException, CanNotCallChefException, EntityNotFoundException, ClientProtocolException, IOException, KeyManagementException, NoSuchAlgorithmException {
```
Comment (flopezag) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/manager/NodeManagerImplTest.java]: gt 120

File: core/src/test/java/com/telefonica/euro_iaas/sdc/manager/NodeManagerImplTest.java [hunk 22]
```diff
@@ -98,22 +106,18 @@ public void deleteNodeTestOK() throws NodeExecutionException, CanNotCallChefExce
 
         when(productInstanceDao.findByHostname(anyString())).thenReturn(productInstances);
        
-        HttpResponse httpResponse= mock(HttpResponse.class);
-        StatusLine statusLine= mock(StatusLine.class);
-        when(httpResponse.getStatusLine()).thenReturn(statusLine);
-        when(httpResponse.getStatusLine().getStatusCode()).thenReturn(200);
-        when(client.execute((HttpUriRequest) anyObject())).thenReturn(httpResponse);
+        when(httpsClient.doHttpsDelete(Mockito.anyString(), Mockito.anyString(), (Map<String,String>)Mockito.anyObject())).thenReturn(200);
```
Comment (flopezag) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/manager/NodeManagerImplTest.java]: gt 120

File: core/src/test/java/com/telefonica/euro_iaas/sdc/manager/NodeManagerImplTest.java [hunk 23]
```diff
@@ -98,22 +106,18 @@ public void deleteNodeTestOK() throws NodeExecutionException, CanNotCallChefExce
...
         
         when(propertiesProvider.getProperty(anyString())).thenReturn("URL");
 
         nodeManager.nodeDelete("test", "testOk", "token");
         
-        verify(httpResponse,times(2)).getStatusLine();
-        verify(client, times(1)).execute((HttpUriRequest) anyObject());
+        verify(httpsClient,times(1)).doHttpsDelete(Mockito.anyString(), Mockito.anyString(), (Map<String,String>)Mockito.anyObject());
```
Comment (flopezag) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/manager/NodeManagerImplTest.java]: idem

File: core/src/test/java/com/telefonica/euro_iaas/sdc/manager/NodeManagerImplTest.java [hunk 24]
```diff
@@ -98,22 +106,18 @@ public void deleteNodeTestOK() throws NodeExecutionException, CanNotCallChefExce
...
+//        verify(client, times(1)).execute((HttpUriRequest) anyObject());
     }
     
     @Test
-    public void deleteNodeTestOK_nodeNotFountInPuppet() throws NodeExecutionException, CanNotCallChefException, EntityNotFoundException, ClientProtocolException, IOException {
+    public void deleteNodeTestOK_nodeNotFountInPuppet() throws NodeExecutionException, CanNotCallChefException, EntityNotFoundException, ClientProtocolException, IOException, KeyManagementException, NoSuchAlgorithmException {
```
Comment (flopezag) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/manager/NodeManagerImplTest.java]: idem

File: core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+        };
+        // Install the all-trusting host verifier
+        HttpsURLConnection.setDefaultHostnameVerifier(allHostsValid);
+        HttpsURLConnection con = (HttpsURLConnection) url.openConnection();
+       
+        return con;
+        
+    }
+
+    public int doHttpsPost(String _url, String payload, Map<String, String> headers) throws KeyManagementException, NoSuchAlgorithmException, IOException {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java]: gt 120

File: core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java [hunk 26]
```diff
@@ -0,0 +1,167 @@
...
+    
+    private HttpsURLConnection connectionSetup(URL url) throws NoSuchAlgorithmException, KeyManagementException, IOException{
+     // Install the all-trusting trust manager
+        SSLContext sslctx = SSLContext.getInstance("SSL");
+        javax.net.ssl.TrustManager[] trustAllCerts = new TrustManager[] { new X509TrustManager() {
+            public X509Certificate[] getAcceptedIssuers() {
+                return null;
+            }
+
+            public void checkClientTrusted(X509Certificate[] certs, String authType) {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/util/HttpsClient.java]: coding style...

File: core/src/main/java/com/telefonica/euro_iaas/sdc/manager/impl/NodeManagerImpl.java (hunk 7 above)
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/sdc/manager/impl/NodeManagerImpl.java]: NTC

File: core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java (hunk 14 above)
Comment (alberts-tid) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java]: Done

Comment (flopezag) [inline core/src/test/java/com/telefonica/euro_iaas/sdc/installator/InstallatorPuppetTest.java]: Done?

**Comment by jframos** [comment]
LGTM

**Pull request merged**